gh-140746: Fix Thread.start() that can hang indefinitely

ryv-odoo opened the pull request:
When a new thread is started (`Thread.start()`), the current thread waits for the new thread's `_started` signal (`self._started.wait()`). If the new thread doesn't have enough memory, it might crash before signaling its start to the parent thread, causing the parent thread to wait indefinitely (still in `Thread.start()`).

To fix the issue, remove the _started python attribute from the Thread class and converted the logic at C level (PyEvent). A flaw of this method, it that the threading module will still contains the zombie thread into the _limbo dictionnary.

We also change `Thread._delete()` to use `pop` to remove the thread from `_active`, as there is no guarantee that the thread exists in `_active[get_ident()]`, thus avoiding a potential `KeyError`. This can happen if `_bootstrap_inner` crashes before `_active[self._ident] = self` executes. We use `self._ident` because we know `set_ident()` has already been called.

Moreover, remove the old comment in `_delete` because `_active_limbo_lock` became reentrant in commit 243fd01047ddce1a7eb0f99a49732d123e942c63.

_Not sure if this fix need/can to be backported_

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-140746 -->
* Issue: gh-140746
<!-- /gh-issue-number -->

Commit:
gh-140746: Fix Thread.start() can hang forever

When a new thread is started (`Thread.start()`), the current thread
waits for the new thread's `_started` signal (`self._started.wait()`).
If the new thread doesn't have enough memory, it might crash before
signaling its start to the parent thread, causing the parent thread
to wait indefinitely (still in `Thread.start()`).

To fix the issue, remove the _started python attribute from the
Thread class and moved the logic on the _os_thread_handle.
WIP

We also change `Thread._delete()` to use `pop` to remove the thread from
`_active`, as there is no guarantee that the thread exists in
`_active[get_ident()]`, thus avoiding a potential `KeyError`.
This can happen if `_bootstrap_inner` crashes before
`_active[self._ident] = self` executes. We use `self._ident` because we
know `set_ident()` has already been called.

Moreover, remove the old comment in `_delete` because
`_active_limbo_lock` became reentrant in commit 243fd01.

YvesDup commented on Modules/_threadmodule.c:
```diff
@@ -359,6 +360,8 @@ thread_run(void *boot_raw)
     // Wait until the handle is marked as running
     PyEvent_Wait(&boot->handle_ready);
+    // Signal that this Thread is running
+    _PyEvent_Notify(&boot->handle_running);
```
Reference to your issue description (**This MemoryError can occur at the C level during the `PyObject_Call` to the `_bootstrap` method**), It might be better to move this statement in the successful block of the call to `PyObject_Call`.

YvesDup commented on Lib/test/test_threading.py:
```diff
@@ -1457,6 +1458,32 @@ def run_in_bg():
+        serving_thread.join(0.1)
+        self.assertFalse(serving_thread.is_alive())
+
+
```
Is that worth testing a use case where the _`bootstrap` method will be never called just to check that the thread is not running ? 
Is that worth testing a use case where the an exception is raised in the `_start_joinable_thread` just to check that the thread is not running ? This last one could simulate your initial issue.

ZeroIntensity commented:
The `_Py` prefix is supposed to denote that an API is private, so I'd guess that there was originally some plan to make `PyEvent` public, and it just never happened.

Commit:
Fix test:
- Redo the change done in test_various_ops because I don't want to change the semantic.
- Add a check in test_memory_error_bootstrap to ensure no dangling thread remains.

Commit:
Fixes of my previous try:

- Add a new type of ThreadHandleState for failing bootstrap (avoid
using THREAD_HANDLE_DONE for two different case, which is incorrect
for join())
- is_bootstraped instead of is_running having the same behavior than
before (True when the thread is `THREAD_HANDLE_DONE` state)
- Renamming stuff

Commit:
- Linting
- Remove useless change
- Inline method ThreadHandle_set_bootstraped

Commit:
- Improve comments
- Useless signaling

ZeroIntensity commented:
The sanitizer check is our fault, it should be fixed on main.

ZeroIntensity committed:
Merge branch 'main' into main-fix-140746-threading-start-waiting

ryv-odoo commented:
Hello @ZeroIntensity ,

> The sanitizer check is our fault, it should be fixed on main.

Ho, thank you for the info and for updating the PR.
I have made the requested changes; please review again

Commit:
Merge branch 'main' into main-fix-140746-threading-start-waiting

Commit:
Fix "boostraped" spelling and comments

ryv-odoo committed:
Merge branch 'main' into main-fix-140746-threading-start-waiting

Commit:
- rename bootstrapped => running
- merge wait method + is_failed: wait return False if the boostrapping has failed.

ryv-odoo committed:
Merge branch 'main' into main-fix-140746-threading-start-waiting

ryv-odoo committed:
Merge branch 'main' into main-fix-140746-threading-start-waiting

ryv-odoo committed:
Merge branch 'main' into main-fix-140746-threading-start-waiting

ryv-odoo committed:
Merge branch 'main' into main-fix-140746-threading-start-waiting

ryv-odoo commented:
Hi @ZeroIntensity , just a gentle reminder on this one.

Is there any specific part of the logic I can clarify or provide more context on to help with the review?

Thanks again for your time! Have a nice day.

ZeroIntensity commented:
Sorry about the delay. I plan to review this during the PyCon sprints on Monday.

ryv-odoo committed:
Merge branch 'main' into main-fix-140746-threading-start-waiting

reneleonhardt commented:
@ryv-odoo Thank your contributing a fix!

Is one positive test case enough?
